### DISCUSSION TITLE
[Need help/Question] How to write ssh-agent service

### DISCUSSION THREAD

**Discussion by TornaxO7** [davmac314/dinit#106] [Q&A] [answered]
Hello! I'm trying to write a  `ssh-agent` service file which should start a ssh-agent for me.
I've the following file at the moment:
`~/.config/dinit.d/ssh-agent`:
```
type = process
command = /usr/bin/ssh-agent -D -a ~/.cache/ssh/ssh-agent.socket
restart = true
```
but if I do `dinitctl start ssh-agent` I'm seeting the following messages:
```
dinit: Service ssh-agent process terminated with exit code 1
[STOPPD] ssh-agent
[  OK  ] ssh-agent
dinit: Service ssh-agent process terminated with exit code 1
[STOPPD] ssh-agent
[  OK  ] ssh-agent
dinit: Service ssh-agent process terminated with exit code 1
[STOPPD] ssh-agent
[  OK  ] ssh-agent
dinit: Service ssh-agent process terminated with exit code 1
[STOPPD] ssh-agent
dinit: Service ssh-agent restarting too quickly; stopping.
[FAILED] ssh-agent
```
do you have an idea what I'm doing wrong?

**Comment by TornaxO7** [comment]
Sadly `ssh-agent` doesn't provide an error message (output from `dinitctl status ssh-agent`):
```
Service: ssh-agent
    State: STOPPED (terminated; exited - status 1)
```

**Comment by mobin-2008** [comment]
Hi.
It looks like you want to use dinit as user service manager.
Please check these few:
1. Run command in shell & check result
2. set `logfile = log-file-path` in your service (Example: `logfile = ~/templog`) then restart it & check your logfile

I converted your issue to discussion; because its better place to talk about your issue (For example you can set a reply as answer).

**Comment by TornaxO7** [comment]
Ok, so I changed `~/.config/dinit.d/ssh-agent` to:
```
type = process
command = eval "$(ssh-agent)"
restart = true
logfile = ~/.cache/ssh/log.log
```
since doing `eval "$(ssh-agent)"` works if I execute it manually. But running `dinitctl start ssh-agent` gives me the following:
```
Service 'ssh-agent' failed to start.
```
Running `dinitctl status ssh-agent` returns the following:
```
Service: ssh-agent
    State: STOPPED (could not be launched)
        Stage: setting up standard input/output descriptors
        Error: No such file or directory
```
I'm a little bit irritated by this o.O

**Reply by mobin-2008** [reply]
Because if you want set logfile on `~/.cache/ssh/log.log` you need to create all needed folders (dinit dont create them) i think you dont have `ssh` folder on `~/.cache`; create it & test it agian

**Reply by TornaxO7** [reply]
I do have `~/.cache/ssh` but it still doesn't work.

**Reply by mobin-2008** [reply]
make sure you have `~/.cache/ssh` after if dont work try `dinitctl reload ssh-agent`

**Reply by TornaxO7** [reply]
Hm... it works fine now but it tops right away. I think that `ssh-agent` is supposed to be in a service :thinking: I'll look for another solution but thank your for your time!

**Reply by mobin-2008** [reply]
What do you mean?
I think you mistake to choose right type for your service. `ssh-agent` seems like one-time by default; if you need `ssh-agent` to be supervised by dinit; run `ssh-agent` with `-D` flag; from `ssh-agent` man page:
>      -D      Foreground mode.  When this option is specified, ssh-agent will not fork

and everythink must be ok

**Reply by mobin-2008** [reply]
@TornaxO7 Have a nice day/night :)

**Reply by TornaxO7** [reply]
> @TornaxO7 Have a nice day/night :)

Thank you, you too c(^-^)c

**Comment by davmac314** [comment]
`type = process` should be used for processes which are long-lived; ssh-agent in its default mode forks and the original process terminates, which is why the service stops straight away. The `-D` command line argument should prevent this, as mobin-2008 suggested. However, you need a way to capture the environment variables (particularly SSH_AUTH_SOCK) that it outputs.

One way to do that would be to have ssh-agent run a script which then stores the relevant environment variables back in dinit's environment (via `dinitctl setenv`). ssh-agent accepts a program to run as an argument.

The reason that `command = eval "$(ssh-agent)"` doesn't work is because the command is not a shell command and doesn't support the `$( ... )` subshell syntax.

**Reply by davmac314** [reply]
Actually running a script under ssh-agent doesn't work as well as it might since ssh-agent will terminate when the script does. With that in mind I would instead suggest running as a `type = scripted` service, with something like (untested):

```
type = scripted
command = ssh-agent-script.sh start
stop-command = ssh-agent-script.sh stop
```

and in ssh-agent-script.sh:

```
#!/bin/sh
if [ "$1" == "start" ]; then
    eval "$(ssh-agent)"
    dinitctl setenv SSH_AUTH_SOCK SSH_AGENT_PID
elsif [ "$1" == "stop" ]; then
    ssh-agent -k
    dinitctl setenv SSH_AUTH_SOCK="" SSH_AGENT_PID=""
fi
```

**Reply by TornaxO7** [reply]
That's an interesting idea, I'll look what I can do, thank you for this information!

**Comment by Duncaen** [comment, marked as answer]
I just use a static location for the socket (`-a` flag) so I can just export the variable from my shell profile, I don't think there is any value in the "tmp" path unless you run multiple ssh-agents.

**Reply by TornaxO7** [reply]
Could you please eloberate your setup?

**Reply by Duncaen** [reply]
Basically the same as you initial post, the problem with yours is `~` which doesn't get expanded since the command doesn't run in a shell.
You can use `$HOME` and enable `load-options  =  sub-vars`.

```
type = process
load-options = sub-vars
command = /usr/bin/ssh-agent -D -a $HOME/.cache/ssh/ssh-agent.socket
```

**Reply by TornaxO7** [reply]
The problem is that I'm getting the following error message if I use the following:
`~/.config/dinit.d/ssh-agent`
```
type = process
load-options = sub-vars
command =/usr/bin/ssh-agent -D -a $HOME/.cache/ssh/ssh-agent.socket
logfile = ~/.cache/ssh/log.log
```
Then executing `dinitctl start ssh-agent`; Then output from `dinitctl start ssh-agent`:
```
Service: ssh-agent
    State: STOPPED (could not be launched)
        Stage: setting up standard input/output descriptors
        Error: No such file or directory
```

**Reply by Duncaen** [reply]
>         Stage: setting up standard input/output descriptors

`logfile` has the same "issue", it doesn't expand `~`, use `$HOME`:

```
type = process
load-options = sub-vars
command = /usr/bin/ssh-agent -D -a $HOME/.cache/ssh/ssh-agent.socket
logfile = $HOME/.cache/ssh/log.log
```

**Reply by TornaxO7** [reply]
Sadly I'm still getting the same error with the following config:
`~/.config/dinit.d/ssh-agent`
```
type = process
load-options = sub-vars
command = /usr/bin/ssh-agent -D -a $HOME/.cache/ssh/ssh-agent.socket
logfile = $HOME/.cache/ssh/log.log
```
:(

**Reply by Duncaen** [reply]
Does `$HOME/.cache/ssh` exist? Have you unloaded, the service? (not sure if reloading is enough)

**Reply by TornaxO7** [reply]
> Does $HOME/.cache/ssh exist?

Yes, output of `pwd`:
```
/home/tornax/.cache/ssh
```

> Have you unloaded, the service?

I only started it with `dinitctl start ssh-agent` because it always "crashed"/stopped due to the error above if I'm right, because it's saying:
```
Service 'ssh-agent' failed to start.
```
if I run `dinitctl start ssh-agent`.

**Reply by Duncaen** [reply]
```
$ dinitctl unload ssh-agent
$ dinitctl start ssh-agent
```

**Reply by TornaxO7** [reply]
Oh, the service is now working but I'm still getting the error:
```
Error connecting to agent: No such file or directory
```
if I use `ssh-add` :thinking:

**Reply by Duncaen** [reply]
You have to export `SSH_AUTH_SOCK=$HOME/.cache/ssh/ssh-agent.socket` in your shells environment.

**Reply by TornaxO7** [reply]
Yay! It's working now! Thank you sir :)
